optaplanner-examples/src/main/java/org/optaplanner/examples/tennis/optional/score/TennisConstraintProvider.java

```diff
@@ -0,0 +1,80 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaplanner.examples.tennis.optional.score;
+
+import org.optaplanner.core.api.score.buildin.hardmediumsoft.HardMediumSoftScore;
+import org.optaplanner.core.api.score.stream.Constraint;
+import org.optaplanner.core.api.score.stream.ConstraintFactory;
+import org.optaplanner.core.api.score.stream.ConstraintProvider;
+import org.optaplanner.core.api.score.stream.Joiners;
+import org.optaplanner.examples.tennis.domain.TeamAssignment;
+import org.optaplanner.examples.tennis.domain.UnavailabilityPenalty;
+
+import static org.optaplanner.core.api.score.stream.ConstraintCollectors.loadBalanceByCount;
+
+public class TennisConstraintProvider implements ConstraintProvider {
+
+    @Override
+    public Constraint[] defineConstraints(ConstraintFactory constraintFactory) {
+        return new Constraint[] {
+                oneAssignmentPerDatePerTeam(constraintFactory),
+                unavailabilityPenalty(constraintFactory)
+        };
+    }
+
+    // ############################################################################
+    // Hard constraints
+    // ############################################################################
+
+    private Constraint oneAssignmentPerDatePerTeam(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(TeamAssignment.class)
+                .ifExistsOther(TeamAssignment.class,
+                        Joiners.equal(TeamAssignment::getTeam),
+                        Joiners.equal(TeamAssignment::getDay),
+                        Joiners.lessThan(TeamAssignment::getId),
+                        Joiners.filtering((teamAssignment, otherTeamAssignment) -> teamAssignment.getTeam() != null))
+                .penalize("oneAssignmentPerDatePerTeam", HardMediumSoftScore.ONE_HARD);
+    }
+
+    private Constraint unavailabilityPenalty(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(UnavailabilityPenalty.class)
+                .ifExists(TeamAssignment.class,
+                          Joiners.equal(UnavailabilityPenalty::getTeam, TeamAssignment::getTeam),
+                          Joiners.equal(UnavailabilityPenalty::getDay, TeamAssignment::getDay))
+                .penalize("unavailabilityPenalty", HardMediumSoftScore.ONE_HARD);
+    }
+
+    // ############################################################################
+    // Medium constraints
+    // ############################################################################
+
+    // Faster, but does not combine well with other constraints in the same level
+    private Constraint fairAssignmentCountPerTeam(ConstraintFactory constraintFactory) {
+
+        //return (long) (Math.sqrt((double) squaredSum) * 1000);
+        return constraintFactory.from(TeamAssignment.class)
+                .filter(teamAssignment -> teamAssignment.getTeam() != null)
```

Comment:
This shouldn't be necessary, I think. `from()` in CS gives you the entities only after they've been already initialized.

Reply:
yep, fromUninitialized() is the alternative that doesn't do that (but don't use that here)